limacharlie/hive.go

```diff
@@ -0,0 +1,268 @@
+		return nil, err
+	}
+
+	// remove data field from return set
+	for k, v := range hiveSet {
+		hiveSet[k] = HiveData{SysMtd: v.SysMtd, UsrMtd: v.UsrMtd}
+	}
+
+	if isPrint {
+		h.printData(hiveSet)
+	}
+
+	return hiveSet, nil
+}
+
+func (h *HiveClient) Get(args HiveArgs, isPrint bool) (*HiveData, error) {
+	if args.Key == "" {
+		return nil, errors.New("key is required")
+	}
+
+	var hiveSet HiveData
+	if err := h.Organization.client.reliableRequest(http.MethodGet,
+		fmt.Sprintf("hive/%s/%s/%s/data", args.HiveName, args.PartitionKey, args.Key), makeDefaultRequest(&hiveSet)); err != nil {
+		return nil, err
+	}
+
+	if isPrint {
+		h.printData(hiveSet)
+	}
+
+	return &hiveSet, nil
+}
+
+func (h *HiveClient) GetMTD(args HiveArgs, isPrint bool) (*HiveData, error) {
+	if args.Key == "" {
+		return nil, errors.New("key is required")
+	}
+
+	var hd HiveData
+	if err := h.Organization.client.reliableRequest(http.MethodGet,
+		fmt.Sprintf("hive/%s/%s/%s/mtd", args.HiveName, args.PartitionKey, args.Key), makeDefaultRequest(&hd)); err != nil {
+		return nil, err
+	}
+	hd.Data = nil
+
+	if isPrint {
+		h.printData(hd)
+	}
+
+	return &hd, nil
+}
+
+func (h *HiveClient) Add(args HiveArgs, isPrint bool) (*HiveResp, error) {
+	if args.Key == "" {
+		return nil, errors.New("key required")
+	}
+
+	target := "mtd" // if no data set default to target type mtd
+	var data map[string]interface{}
+	if args.Data != nil {
+		// ensure passed data can unmarshal correctly
+		err := json.Unmarshal(args.Data, &data)
+		if err != nil {
+			return nil, err
+		}
+		target = "data"
+	}
+
+	var userMtd UsrMtd // set UsrMtd Data
+	if args.Expiry != 0 {
+		userMtd.Expiry = args.Expiry
+	}
+	if args.Enabled {
+		userMtd.Enabled = args.Enabled
+	}
+	if len(args.Tags) != 0 {
+		userMtd.Tags = args.Tags
+	}
+
+	reqDict := Dict{
+		"data":    data,
+		"usr_mtd": userMtd,
+	}
+
+	if args.ETag != "" {
+		reqDict["etag"] = args.ETag
+	}
+
+	var hiveResp HiveResp
+	req := makeDefaultRequest(&hiveResp).withQueryData(reqDict)
+	if err := h.Organization.client.reliableRequest(http.MethodPost,
+		fmt.Sprintf("hive/%s/%s/%s/%s", args.HiveName, args.PartitionKey, args.Key, target), req); err != nil {
+		return nil, err
+	}
+
+	if isPrint {
+		h.printData(hiveResp)
+	}
+
+	return &hiveResp, nil
+}
+
+func (h *HiveClient) Update(args HiveArgs, isPrint bool) (interface{}, error) {
+	if args.Key == "" {
+		return nil, errors.New("key required")
+	}
+
+	target := "mtd" // if no data set default to target type mtd
+	var existing *HiveData
+	var err error
+	if args.Data != nil {
+		var data map[string]interface{}
+		err = json.Unmarshal(args.Data, &data)
+		if err != nil {
+			return nil, err
+		}
+		target = "data"
+
+		existing, err = h.Get(args, false)
+		if err != nil {
+			return nil, err
+		}
+		existing.Data = data
+	} else {
+		existing, err = h.GetMTD(args, false)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// set usr mtd data
+	if args.Expiry != 0 {
```

Comment:
I think we should update the args to be pointers to the data. Right now, if I want to update an existing record and remove its expiry, I would specify `0` but we interpret that as "not set", which means I could never call this function to:
- disable expiry
- disable a record
- remove all tags

But if we make `Enabled *bool` for example, it would allow us to make a distinction between "0" and "don't change" (`nil`).

Reply:
Good catch, updated all params to pointers